client.go

```diff
@@ -129,106 +154,144 @@ func (c *Client) FetchKeyQuota(ctx context.Context, accessKey, origin string, no
+		}
+		if perm != proto.UserPermission_UNAUTHORIZED {
+			return perm, access, nil
 		}
 	}
 
 	// Ask quotacontrol server via client
-	if userPerm == nil {
-		userPerm, resourceAccess, err = c.quotaClient.GetUserPermission(ctx, projectID, userID)
-		if err != nil {
-			return userPerm, resourceAccess, err
-		}
+	perm, access, err := c.quotaClient.GetUserPermission(ctx, projectID, userID)
+	if err != nil {
+		logger.Error("unexpected client error", slog.Any("err", err))
+		return proto.UserPermission_UNAUTHORIZED, nil, err
 	}
+	return perm, access, nil
+}
 
-	// Check if userPerm is still nil, in which case return unauthorized
-	if userPerm == nil {
-		v := proto.UserPermission_UNAUTHORIZED
-		return &v, resourceAccess, nil
+func (c *Client) SpendQuota(ctx context.Context, quota *proto.AccessQuota, computeUnits int64, now time.Time) (spent bool, total int64, err error) {
+	// quota is nil only on unexpected errors from quota fetch
+	if quota == nil || computeUnits == 0 {
+		return false, 0, nil
 	}
 
-	return userPerm, resourceAccess, nil
-}
+	logger := c.logger.With(
+		slog.String("op", "spend_quota"),
+		slog.Uint64("project_id", quota.AccessKey.ProjectID),
+		slog.String("access_key", quota.AccessKey.AccessKey),
+	)
 
-func (c *Client) SpendQuota(ctx context.Context, quota *proto.AccessQuota, computeUnits int64, now time.Time) (bool, error) {
 	accessKey := quota.AccessKey.AccessKey
 	cfg := quota.Limit
 
 	// spend compute units
 	key := getQuotaKey(quota.AccessKey.ProjectID, quota.Cycle, now)
 
-	for i := time.Duration(0); i < 3; i++ {
+	for i := range 3 {
 		total, err := c.usageCache.SpendComputeUnits(ctx, key, computeUnits, cfg.OverMax)
-		switch err {
-		case nil:
-			usage, event := cfg.GetSpendResult(computeUnits, total)
-			if quota.AccessKey.AccessKey == "" {
-				c.usage.AddProjectUsage(quota.AccessKey.ProjectID, now, usage)
-			} else {
-				c.usage.AddKeyUsage(accessKey, now, usage)
-			}
-			if usage.LimitedCompute != 0 {
-				return false, proto.ErrLimitExceeded
+		if err != nil {
+			// limit exceeded
+			if errors.Is(err, proto.ErrLimitExceeded) {
+				c.usage.AddKeyUsage(accessKey, now, proto.AccessUsage{LimitedCompute: computeUnits})
+				return false, total, proto.ErrLimitExceeded
 			}
-			if event != nil {
-				if _, err := c.quotaClient.NotifyEvent(ctx, quota.AccessKey.ProjectID, event); err != nil {
-					c.logger.With("err", err, "op", "use_access_key", "event", event).Error("-> quota control: failed to notify")
+			// ping the server to prepare usage
+			if errors.Is(err, ErrCachePing) {
+				if _, err := c.quotaClient.PrepareUsage(ctx, quota.AccessKey.ProjectID, quota.Cycle, now); err != nil {
+					logger.Error("unexpected client error", slog.Any("err", err))
+					if _, err := c.usageCache.ClearComputeUnits(ctx, key); err != nil {
+						logger.Error("unexpected cache error", slog.Any("err", err))
+					}
+					return false, 0, nil
 				}
+				continue
```

Comment:
should we sleep after the ping? or do you expect SpendComputeUnits to be available right away?